[REBASE & FF] Clean Up Noisy TCG Prints

os-d opened the pull request:
## Description

This PR consists of three commits:

SecurityPkg: Remove/Downgrade Noisy TCG Prints
--
The TCG code is very noisy when a TPM is connected. This commit downgrades some prints to verbose and removes some others that do not have value (such as function enter and exit prints).

SecurityPkg: Tpm2DebugLib: Clean Up Foundations
--
The Tpm2DebugLib varieties were missing some common things, such as including their own header, SecurityPkg.dec as a package, and including headers for data structures used in the source code.

SecurityPkg: Tpm2DebugLib: Move Noisy Debug Prints to Verbose Lib
--
The TCG code is very noisy, especially when printing PCR digests and the event log. E.g. in a platform printing at info level, ~4000 messages of the 5700 messages in the serial log were from the TCG code.

This patch moves the noisy prints to the verbose version of Tpm2DebugLib so that the log is not spammed, the code and strings can be dropped from the FW, but it is still available for debugging when needed.

Note, the definition of a noisy log is subjective, so please feel free to note the usefulness of a given log and it can be kept or modified.

Note pt 2: This entire library will be upstreamed to edk2 after these changes are merged.

Note pt 3: `DUMP_HEX` is also not upstreamed to edk2, so after this library is upstreamed with the direct copy, the custom hex dumping will be removed in favor of `DUMP_HEX` (along with other places in the code using custom hex dumping).

- [ ] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [x] Backport to release branch?

## How This Was Tested

Build tested.

## Integration Instructions

Ensure Tpm2DebugLib is in the `[LibraryClasses]` section of the platform DSC. It is anticipated it already is due to its use in Tcg2Dxe, which is why this breaking change is backported to the release branch; it is not expected to break platforms currently using Mu.

os-d committed:
SecurityPkg: Remove/Downgrade Noisy TCG Prints

The TCG code is very noisy when a TPM is connected. This
commit downgrades some prints to verbose and removes some
others that do not have value (such as function enter and
exit prints).

Signed-off-by: Oliver Smith-Denny <osde@microsoft.com>

os-d committed:
SecurityPkg: Tpm2DebugLib: Clean Up Foundations

The Tpm2DebugLib varieties were missing some common things, such
as including their own header, SecurityPkg.dec as a package,
and including headers for data structures used in the source code.

Signed-off-by: Oliver Smith-Denny <osde@microsoft.com>

os-d committed:
SecurityPkg: Tpm2DebugLib: Move Noisy Debug Prints to Verbose Lib

The TCG code is very noisy, especially when printing PCR digests
and the event log. E.g. in a platform printing at info level,
~4000 messages of the 5700 messages in the serial log were from
the TCG code.

This patch moves the noisy prints to the verbose version of
Tpm2DebugLib so that the log is not spammed, the code and strings
can be dropped from the FW, but it is still available for debugging
when needed. The debug logic is simply moved, not edited. A following
commit updates the debug logic for spelling and functionality.

Signed-off-by: Oliver Smith-Denny <osde@microsoft.com>

spbrogan commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -9,10 +9,14 @@ MU_CHANGE
+#include <Library/Tpm2DebugLib.h>
 
 #include <IndustryStandard/Tpm20.h>
+#include <IndustryStandard/UefiTcgPlatform.h>
```
we may need to revert a recent PR which we have decided is not needed which moved more things to this header file.  Might want to check with raymond

os-d replied:
Will do

os-d replied:
I checked with Raymond, we should not need to structure the removal of the additional items in the header file first, I am not using anything new. This header file was added here because of some existing definitions that the code from Tcg2Dxe was using to do the dumping, so it was needed here as well, but again, none of the new definitions are being used, so this is unaffected by the revert (or if it is not reverted).

os-d replied:
Raymond's change would have affected this PR, but it went into 202405 and has not been cherry-picked to 202502 yet (and sounds like it won't be). So the headers here are correct.

apop5 commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+
+  DEBUG ((DEBUG_INFO, "\n"));
+  DEBUG ((DEBUG_INFO, "    EventSize - 0x%08x\n", EventHdr->EventSize));
+  InternalDumpHex ((UINT8 *)(EventHdr + 1), EventHdr->EventSize);
```
Not using DUMP_HEX because you are upstreaming this? https://github.com/microsoft/mu_basecore/blob/d3d9be22570f80ea322c87ea0bb16c036778f793/MdePkg/Include/Library/DebugLib.h#L689

os-d replied:
Correct, that's the note I left in the PR text, that I am doing just the straight copy over to upstream this first, then plan to upstream DUMP_HEX and change this (and all other locations with custom hex dumping code) after that. I think it will be an easier upstream process.

makubacki commented on SecurityPkg/Include/Library/Tpm2DebugLib.h:
```diff
@@ -40,4 +42,37 @@ DumpTpmOutputBlock (
+
+  @param[in]  PcrIndex   The index of the PCR.
+  @param[in]  HashAlg    The hash algorithm used.
+  @param[in]  Digest     The digest to be dumped.
```
```suggestion
  @param[in]  PcrValues  The digest to be dumped.
```

os-d replied:
Thanks, I will go through and fix all of these. For the ones that were copied from the original code, I will do a follow on commit to address, so there is a commit that just moves the logic, then a commit that updates it. I think that is both easier to track what is going on and for upstreaming to edk2.

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibNull.c:
```diff
@@ -41,3 +44,42 @@ DumpTpmOutputBlock (
+
+  @param[in]  PcrIndex   The index of the PCR.
+  @param[in]  HashAlg    The hash algorithm used.
+  @param[in]  Digest     The digest to be dumped.
```
```suggestion
  @param[in]  PcrValues  The digest to be dumped.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibSimple.c:
```diff
@@ -78,3 +80,52 @@ DumpTpmOutputBlock (
+
+  @param[in]  PcrIndex   The index of the PCR.
+  @param[in]  HashAlg    The hash algorithm used.
+  @param[in]  Digest     The digest to be dumped.
```
```suggestion
  @param[in]  PcrValues   The digest to be dumped.
```
```suggestion
  @param[in]  Digest     The digest to be dumped.

```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibSimple.c:
```diff
@@ -78,3 +80,52 @@ DumpTpmOutputBlock (
+  // Dumping the entire event log is left for the verbose version of this library, because it can be thousands of
+  // lines long.
+  DEBUG ((DEBUG_INFO, "EventLogFormat - 0x%08x\n", EventLogFormat));
+  DEBUG ((DEBUG_INFO, "EventLogLocation - 0x%08x\n", EventLogLocation));
+  DEBUG ((DEBUG_INFO, "EventLogLastEntry - 0x%08x\n", EventLogLastEntry));
+  DEBUG ((DEBUG_INFO, "FinalEventsTable - 0x%08x\n", FinalEventsTable));
```
Is there a reason to not just print the addresses as 64-bit values?

os-d replied:
Because tab autocomplete from Copilot did this :). I will fix it up.

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibSimple.c:
```diff
@@ -78,3 +80,52 @@ DumpTpmOutputBlock (
 } // DumpTpmOutputBlock()
+
+/**
+  This function dump event log.
```
```suggestion
  This function dumps the event log.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibSimple.c:
```diff
@@ -78,3 +80,52 @@ DumpTpmOutputBlock (
+  DEBUG ((DEBUG_INFO, "EventLogLocation - 0x%08x\n", EventLogLocation));
+  DEBUG ((DEBUG_INFO, "EventLogLastEntry - 0x%08x\n", EventLogLastEntry));
+  DEBUG ((DEBUG_INFO, "FinalEventsTable - 0x%08x\n", FinalEventsTable));
+
+  return;
```
```suggestion
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibSimple.c:
```diff
@@ -78,3 +80,52 @@ DumpTpmOutputBlock (
+  // This function simply marks we have read these PCRs, the verbose version of this function will dump the contents.
+  DEBUG ((DEBUG_INFO, "PcrIndex - %d\n", PcrIndex));
+  DEBUG ((DEBUG_INFO, "HashAlg - 0x%04x\n", HashAlg));
+  return;
```
Feel free to keep these, but I don't see the need.

```suggestion
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
 } // DumpTpmOutputBlock()
+
+/**
+
+  This function dump raw data.
```
```suggestion
  This function dumps raw data.

```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+
+  This function dump raw data.
+
+  @param  Data  raw data
+  @param  Size  raw data size
```
```suggestion
  @param[in]  Data  raw data
  @param[in]  Size  raw data size
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+
+/**
+
+  This function dump raw data with colume format.
```
```suggestion
/**
  This function dumps raw data formatted in coumns.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+
+  This function dump raw data with colume format.
+
+  @param  Data  raw data
+  @param  Size  raw data size
```
```suggestion
  @param  Data[in]  Raw data
  @param  Size[in]  Raw data size

```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+  UINTN  Count;
+  UINTN  Left;
+
+  #define COLUME_SIZE  (16 * 2)
+
+  Count = Size / COLUME_SIZE;
+  Left  = Size % COLUME_SIZE;
+  for (Index = 0; Index < Count; Index++) {
+    DEBUG ((DEBUG_INFO, "%04x: ", Index * COLUME_SIZE));
+    InternalDumpData (Data + Index * COLUME_SIZE, COLUME_SIZE);
+    DEBUG ((DEBUG_INFO, "\n"));
+  }
+
+  if (Left != 0) {
+    DEBUG ((DEBUG_INFO, "%04x: ", Index * COLUME_SIZE));
+    InternalDumpData (Data + Index * COLUME_SIZE, Left);
+    DEBUG ((DEBUG_INFO, "\n"));
+  }
```
```suggestion
  #define COLUMN_SIZE  (16 * 2)

  Count = Size / COLUMN_SIZE;
  Left  = Size % COLUMN_SIZE;
  for (Index = 0; Index < Count; Index++) {
    DEBUG ((DEBUG_INFO, "%04x: ", Index * COLUMN_SIZE));
    InternalDumpData (Data + Index * COLUME_SIZE, COLUMN_SIZE);
    DEBUG ((DEBUG_INFO, "\n"));
  }

  if (Left != 0) {
    DEBUG ((DEBUG_INFO, "%04x: ", Index * COLUMN_SIZE));
    InternalDumpData (Data + Index * COLUMN_SIZE, Left);
    DEBUG ((DEBUG_INFO, "\n"));
  }
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+}
+
+/**
+  This function dump PCR event.
```
```suggestion
  This function dumps PCR events.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+  DEBUG ((DEBUG_INFO, "    PCRIndex  - %d\n", EventHdr->PCRIndex));
+  DEBUG ((DEBUG_INFO, "    EventType - 0x%08x\n", EventHdr->EventType));
+  DEBUG ((DEBUG_INFO, "    Digest    - "));
+  for (Index = 0; Index < sizeof (TCG_DIGEST); Index++) {
```
```suggestion
  for (Index = 0; Index < sizeof (EventHdr->Digest); Index++) {
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+}
+
+/**
+  This function dump PCR event 2.
```
```suggestion
  This function dumps TCG_PCR_EVENT2 type events.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+  DEBUG ((DEBUG_INFO, "\n"));
+  DigestBuffer = DigestBuffer - sizeof (TPMI_ALG_HASH);
+
+  CopyMem (&EventSize, DigestBuffer, sizeof (TcgPcrEvent2->EventSize));
```
I think this would be safer to better associate the max with the actual destination buffer.

```suggestion
  CopyMem (&EventSize, DigestBuffer, sizeof (EventSize));
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+}
+
+/**
+  This function dump TCG_EfiSpecIDEventStruct.
```
```suggestion
  This function dumps a TCG_EfiSpecIDEventStruct structure.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+  This function gets the size of TCG_EfiSpecIDEventStruct. It is copied here as it is a simple calculation and to
+  avoid library dependencies within TCG2 code.
+
+  @param[in]  TcgEfiSpecIdEventStruct     A pointer to TCG_EfiSpecIDEventStruct.
+**/
```
Add return value to function description.

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+  return sizeof (TCG_EfiSpecIDEventStruct) + sizeof (UINT32) + (NumberOfAlgorithms * sizeof (TCG_EfiSpecIdEventAlgorithmSize)) + sizeof (UINT8) + (*VendorInfoSize);
+}
+
+/**
+  This function returns size of TCG PCR event 2.
```
```suggestion
/**
  This function returns the size of a TCG PCR event 2 instance.
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+
+  DigestBuffer = DigestBuffer - sizeof (TPMI_ALG_HASH);
+
+  CopyMem (&EventSize, DigestBuffer, sizeof (TcgPcrEvent2->EventSize));
```
```suggestion
  CopyMem (&EventSize, DigestBuffer, sizeof (EventSize));
```

makubacki commented on SecurityPkg/Library/Tpm2DebugLib/Tpm2DebugLibVerbose.c:
```diff
@@ -736,3 +740,417 @@ DumpTpmOutputBlock (
+}
+
+/**
+  This function dump event log.
```
```suggestion
  This function dumps the event log.
```

os-d committed:
SecurityPkg: Minor Cleanups to Tpm Debug Logic

This commit cleans up some logic in the newly moved TPM
debug logic to fix spelling and a few logical fixes.

Signed-off-by: Oliver Smith-Denny <osde@microsoft.com>
